Add write flattening

tonyhutter opened the pull request:
### Description

This flattens all the writes on an fsync to reduce the number of extents transferred.  

### Motivation and Context
Flattenting writes reduces the number of extents when you're doing lots of overwrites.  For example, let's say you did five writes:
```
[w1---------------]
         [w2-----------]
                [w3-------------]
                [w4-------------]
    [w5----------------]
```
After flattening:
```
[w1][w5----------------][w4-----]
```
This also means that reads are faster, since there's fewer extents to iterate over.  You can disable write flattening via the UNIFYFS_CLIENT_FLATTEN_WRITES config value (enabled by default).

Other notes:

- The red-black tree implementation (tree.h) included in this commit was taken from OpenBSD:
    
http://cvsweb.openbsd.org/cgi-bin/cvsweb/~checkout~/src/sys/sys/tree.h
    
tree.h's macros are valid but checkpatch still flags them.  It's exempted via the commit message:
TEST_CHECKPATCH_SKIP_FILES=common/src/tree.h
    
- Cameron's travis.yml fix is included.

### Motivation and Context
Better performance on overwrite heavy workloads

### How Has This Been Tested?
Tested via manually running the added `examples/src/multi-write.c` example code.  Due to hitting an unrelated corruption bug (https://github.com/LLNL/UnifyFS/issues/411), I've left `multi-write.c` as example code, not a test case.  Once the bug is fixed, we should consider converting it to a test case.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted.

adammoody commented:
Thanks @tonyhutter .  I know this was discussed on a call that I missed, but are we allowed to add LGPL source code to a BSD project?

https://opensource.stackexchange.com/questions/6698/can-i-use-code-from-an-lgpl-v2-licenced-project-in-a-bsd-3-clause-licenced-proje 

If that's a problem, do we have AVL code under BSD?

adammoody commented on client/src/unifyfs-sysio.c:
```diff
@@ -2009,6 +2005,12 @@ int UNIFYFS_WRAP(fsync)(int fd)
 {
     /* check whether we should intercept this file descriptor */
     if (unifyfs_intercept_fd(&fd)) {
+
```
style: kill this blank line after if()

adammoody commented on client/src/unifyfs-sysio.c:
```diff
@@ -998,12 +997,13 @@ ssize_t UNIFYFS_WRAP(read)(int fd, void* buf, size_t count)
     /* check whether we should intercept this file descriptor */
     if (unifyfs_intercept_fd(&fd)) {
+
+        LOGDBG("write %d bytes to fd %d", count, fd);
```
style: swap these two lines, to kill the blank line after if() and then separate LOGDBG statement from following block of code

adammoody commented on client/src/unifyfs.c:
```diff
@@ -914,6 +921,7 @@ int unifyfs_fid_create_file(const char* path)
 {
     /* check that pathname is within bounds */
     size_t pathlen = strlen(path) + 1;
+    int rc;
```
style: let's either declare rc this at the very to of the function followed by a blank line, or move it down into the area where it is used like ```int rc = seg_tree_init(&meta->seg_tree);```

adammoody commented on common/src/seg_tree.c:
```diff
@@ -0,0 +1,311 @@
+/* Returns 0 on success, positive non-zero error code otherwise */
+int seg_tree_init(struct seg_tree* seg_tree)
+{
+	memset(seg_tree, 0, sizeof(*seg_tree));
```
style: extra leading space before memset?

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -543,78 +621,77 @@ static int unifyfs_logio_chunk_write(
+void unifyfs_rewrite_index_from_seg_tree(void)
+{
+    struct seg_tree_node* node;
+    int fid, gfid;
```
style: let's move these three declarations down to the lines where the variables are used:
```
int fid = unifyfs_fds[i].fid;
```

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -471,6 +473,82 @@ static int unifyfs_split_index(
+             * and number remaining */
+            num_entries = 0;
+            *(unifyfs_indices.ptr_num_entries) = num_entries;
+            remaining_entries = unifyfs_max_index_entries - num_entries;
```
Let's we rewrite this code to call your new clear_index function now that we have it

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -543,78 +621,77 @@ static int unifyfs_logio_chunk_write(
+{
+    size_t index_size;
+    index_size = sizeof(*unifyfs_indices.index_entry) *
+        (*unifyfs_indices.ptr_num_entries);
```
style: let's pull (*unifyfs_indices.ptr_num_entries) into a local variable.  This helps a lot when running under a debugger like totalview.
```
    off_t num_entries = *unifyfs_indices.ptr_num_entries;
    size_t index_size = sizeof(*unifyfs_indices.index_entry) * num_entries;
```

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -543,78 +621,77 @@ static int unifyfs_logio_chunk_write(
-        off_t remaining_entries = unifyfs_max_index_entries - num_entries;
+    /* Update our write metadata with the new write */
+    return (unifyfs_logio_add_write_meta_to_index(gfid, pos, log_offset,
+        count));
```
style: let's modify this line to call the function one line and return on another:
```
    int rc = unifyfs_logio_add_write_meta_to_index(gfid, pos, log_offset, count));
    return rc;
```
That makes it easier to see the return value from the function with a debugger.

adammoody commented on common/src/seg_tree.c:
```diff
@@ -0,0 +1,311 @@
+ */
+ /*
+  * This file is a simple, thread-safe, segment tree implementation.  The
+  * segments in the tree are no-overlapping.  Added segments overwrite the old
```
typo? no-overlapping --> non-overlapping

tonyhutter commented:
@adammoody thanks for taking a look at this - let me go though your changes.  I think you bring up a good point with the LGPL.  Looks like I need to change the Makefile to build `rb.c` as a **dynamic** library rather than statically as I have now.  That will make it compliant.

tonyhutter commented:
@MichaelBrim I'll fix the typo, thanks for spotting that.

tonyhutter commented:
> Looks like I need to change the Makefile to build rb.c as a dynamic library rather than statically as
> I have now. That will make it compliant.

Maybe I spoke too soon.  We'd be distributing the modified `rb.c/rb.h` as well, which may be an issue.  Let me look into that.  In the meantime, if you know of a BSD/MIT red-black tree or AVL tree implementation, please let me know.

tonyhutter commented:
@adammoody my latest push swaps out the LGPL'd red black tree implementation in favor of OpenBSD's `tree.h` one. It also fixes all your comments.

I believe the travis errors are due to #415

MichaelBrim commented on scripts/checkpatch.sh:
```diff
@@ -30,6 +30,29 @@ checkpatch_ignore+=",VOLATILE"            # Allow use of volatile keyword
 
 checkpatch_cmd+=" --ignore $checkpatch_ignore"
 
+# SPECIAL CASE: exclude tree.h
```
We have the same sorts of checkpatch failures for the macros in `common/src/unifyfs_configurator.h`. Why have this in the checkpatch.sh script? The whole point of TEST_CHECKPATCH_SKIP_FILES is so you can set it in the commit message. You can just do `TEST_CHECKPATCH_SKIP_FILES=common/src/tree.h` in the commit message.

tonyhutter replied:
I just tried that in the commit message, but it still flagged the errors.

MichaelBrim replied:
Did you do a rebase too?

tonyhutter replied:
I did a `git commit --amend` to re-write the commit message.  I also just did a rebase, but still no luck.

MichaelBrim replied:
weird. I checked out your branch, removed the checkpatch change, and then did:
```
export TEST_CHECKPATCH_SKIP_FILES="common/src/tree.h"
git diff dev | ./scripts/checkpatch.sh
```
It worked for me.

tonyhutter replied:
Yea, it works if you export it as a env var, but doesn't work if you add `TEST_CHECKPATCH_SKIP_FILES=common/src/tree.h` in the commit message.  That's why I had to add it to checkpatch.sh itself, to get it to pass Travis.

MichaelBrim replied:
I also just tried to amend the commit message, and that worked for me too. Wonder what is different in your environment?

tonyhutter replied:
Is `TEST_CHECKPATCH_SKIP_FILES` still exported?  

I don't see anywhere in the code where checkpatch.sh would be looking through the commit message to set env variables.

MichaelBrim replied:
I recreated what travis does by the following commands:
```
> eval $(./scripts/git_log_test_env.sh)
> export TEST_CHECKPATCH_SKIP_FILES
> ./scripts/checkpatch.sh
> echo $TEST_CHECKPATCH_SKIP_FILES
common/src/tree.h
```

MichaelBrim replied:
It's the first command the greps for the TEST_* variables in the commit message and adds them into the environment.

tonyhutter replied:
Ah ok, I wasn't aware Travis doing some magic beforehand when it called checkpatch.  That seems wrong, but that's an issue for another time.  I'll update the commit message with `TEST_CHECKPATCH_SKIP_FILES` set and just export it when I want to checkpatch manually.

MichaelBrim commented on client/src/unifyfs-sysio.c:
```diff
@@ -1708,9 +1707,6 @@ static int process_read_data(read_req_t* read_reqs, int count, int* done)
         req.length  = msg->length;
         req.errcode = msg->errcode;
 
-        LOGDBG("read reply: gfid=%d offset=%zu size=%zu",
```
Why remove this LOGDBG()?

tonyhutter replied:
I can't remember, I'll add it back in.

MichaelBrim commented on client/src/unifyfs-sysio.c:
```diff
@@ -998,7 +997,7 @@ ssize_t UNIFYFS_WRAP(read)(int fd, void* buf, size_t count)
 ssize_t UNIFYFS_WRAP(write)(int fd, const void* buf, size_t count)
 {
-    LOGDBG("write was called for fd %d", fd);
+    LOGDBG("write %d bytes to fd %d", count, fd);
```
should use `%zu` for size_t `count` parameter

tonyhutter replied:
Ok, will fix

adammoody commented:
Here is where write index entries are combined as they are written into the log in the current code:
https://github.com/LLNL/UnifyFS/blob/3f0371ea455fb8d8eb48a1d0c7fcb48a206e24c4/client/src/unifyfs-fixed.c#L568

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -471,6 +472,96 @@ static int unifyfs_split_index(
+        /* if we have filled the key/value buffer, flush it to server */
+        if (0 == remaining_entries) {
+            /* index buffer is full, flush it */
+            int ret = invoke_client_fsync_rpc(cur_idx.gfid);
```
There are a couple places like this where we can invoke the fsync rpc as the user is writing.  There is another one in unifyfs_split_index.  This is when the user has filled up the metadata index.  I suppose we should also rewrite and/or blank out the segment tree at these points.  Do you do that now?

tonyhutter replied:
You bring up a really good issue.  I don't think things will work with my patch with that sync in there.  There may be a way I can reorder things so that the potential "sync on full index"  is done before the seg_tree_add.  Let me look into it and update the patch.

CamStan commented:
@tonyhutter, the issue with Travis was that `AX_PTHREAD` is apparently found in the autoconf archive, which Travis doesn't install by default. I added it to the .travis.yml and pushed it to your branch. It looks like it's working now.

tonyhutter commented:
@CamStan thanks!!!

tonyhutter commented:
@adammoody I pushed changes that address the `unifyfs_coalesce_index()`/`unifyfs_split_index()` issues you pointed out.  However, I broke something else along the way...  let me fix that first and then I'll have you take another look.

tonyhutter commented:
Fixed the issue.  @adammoody can you take another look at this?

adammoody commented:
With the break, I forgot we had this sitting out there.  I'll take a fresh look.

adammoody commented:
Ok, closer.  One concern is whether the rewrite function can overflow the limit of the index array.  I see it has a loop but it doesn't check that there is space.  That could be ok if we know that the combined set of index entries across all trees are not greater than the number of slots, but can we be sure.  If not, we need to fix that.  If so, we should probably add comments to describe that.

In particular, one problem I can see is that the number of entries in the tree for a file could be greater than the number in the index buffer, e.g., if someone were to write a large segment, and then later write a smaller segment that falls entirely within the larger segment.  I think we'd end up with two index entries in the buffer, and three in the tree.  Since we only sync when the buffer fills, the rewrite could then overflow.  Is that right?

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -471,6 +482,146 @@ static int unifyfs_split_index(
+        unifyfs_rewrite_index_from_seg_tree();
+    }
+
+    ret = invoke_client_fsync_rpc(gfid);
```
Let's declare int ret here, and save a couple lines above.

tonyhutter replied:
Ok

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -543,78 +695,64 @@ static int unifyfs_logio_chunk_write(
-    cur_idx.log_pos  = log_offset;
-    cur_idx.length   = count;
+    /* Update our write metadata with the new write */
+    rc = unifyfs_logio_add_write_meta_to_index(meta, pos, log_offset,
```
let's move the type definition of int rc here and drop it from the top of the function, since I don't think we use it anywhere else above.

tonyhutter replied:
Ok

adammoody commented on client/src/unifyfs-fixed.c:
```diff
@@ -471,6 +482,146 @@ static int unifyfs_split_index(
+    off_t num_entries = *unifyfs_indices.ptr_num_entries;
+    size_t index_size = sizeof(*unifyfs_indices.index_entry) * num_entries;
+
+    memset(unifyfs_indices.index_entry, 0, index_size);
```
For performance, we could skip this memset.  Though it would be nice to have in a debug mode.  Can we conditionally do that?

tonyhutter replied:
I'll just get rid of it.  There's really no need for the `memset()`

tonyhutter commented:
> One concern is whether the rewrite function can overflow the limit of the index array. 

That's a good point.  We could simply force a sync if the seg_tree count is greater than or equal to `unifyfs_max_index_entries - 1` (since adding a single write could at most create two seg_tree entries).

adammoody commented:
> > One concern is whether the rewrite function can overflow the limit of the index array.
> 
> That's a good point. We could simply force a sync if the seg_tree count is greater than or equal to `unifyfs_max_index_entries - 1` (since adding a single write could at most create two seg_tree entries).

Yeah, that makes sense.  Since we're flushing all segment trees in one shot, we'd want to use a global count of entries across all active trees.

I think a single write can replace an existing segment tree entry with three, right?  Something like:

existing entry: [50, 100)
new write: [60, 70)
resulting set: [50, 60), [60, 70), [70, 100)

Is that what you're thinking?

tonyhutter commented:
> I think a single write can replace an existing segment tree entry with three, right?

You are correct sir!  I should add in a proper `seg_tree_count()` function too.  So after a write, we'd be doing:
`if seg_tree_count(tree) >= (unifyfs_max_index_entries - 2) then 
    sync()
`.

adammoody commented:
@tonyhutter , the collective work is coming longer faster than I anticipated, and I'm at a point where I could start to you your new range tree in the server.  I see there is a merge conflict at the moment, preventing a merge.  Even if you're not quite done with the final details on this PR, could you fix up the merge conflict?  We can go ahead and merge this and then work in parallel if you're still polishing things up.

tonyhutter commented:
@adammoody will do.  Note that my last push updated the seg_tree with two functions you might need:
```
/* Return the number of segments in the segment tree */
unsigned long seg_tree_count(struct seg_tree* seg_tree)

/* Return the maximum segment value in the tree */
unsigned long seg_tree_max(struct seg_tree* seg_tree)
```

adammoody commented:
> @adammoody will do. Note that my last push updated the seg_tree with two functions you might need:
> 
> ```
> /* Return the number of segments in the segment tree */
> unsigned long seg_tree_count(struct seg_tree* seg_tree)
> 
> /* Return the maximum segment value in the tree */
> unsigned long seg_tree_max(struct seg_tree* seg_tree)
> ```

Cool, we could easily compute the total number of segments for a file with that count function and an allreduce.

tonyhutter commented:
@adammoody you can pull this in - I rebased and fixed the conflict.  I also added code to sync seg_tree if it's going to overflow the index but haven't actually tested it yet.

adammoody commented:
Awesome!  Thanks, @tonyhutter !
